src/hotspot/share/runtime/java.cpp

```diff
@@ -416,6 +417,19 @@ void before_exit(JavaThread* thread, bool halt) {
     }
   }
 
+  // If we are built with LSan, we need to perform leak checking. If we are
+  // terminating normally, not halting and no VM error, we perform a normal
+  // leak check which terminates if leaks are found. If we are not terminating
+  // normally, halting or VM error, we perform a recoverable leak check which
+  // prints leaks but will not terminate.
+  if (!halt && !VMError::is_error_reported()) {
+    LSAN_DO_LEAK_CHECK();
+  } else {
+    int result = LSAN_DO_RECOVERABLE_LEAK_CHECK();
+    // Ignore the return value.
+    static_cast<void>(result);
+  }
+
```

Comment:
Why is this unconditional? I expected it to be in `#ifdef LEAK_SANITIZER`

Reply:
Since this does a little more than just invoke the LSAN_* macros, that is fair. I wrapped it in an ifdef.